bundles/org.openhab.binding.irobot/src/main/java/org/openhab/binding/irobot/discovery/IRobotDiscoveryService.java

```diff
@@ -0,0 +1,168 @@
+import org.openhab.core.config.discovery.DiscoveryResultBuilder;
+import org.openhab.core.config.discovery.DiscoveryService;
+import org.openhab.core.net.NetUtil;
+import org.openhab.core.thing.ThingUID;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonParseException;
+
+/**
+ * Discovery service for iRobot's
+ *
+ * @author Pavel Fedin <pavel_fedin@mail.ru> - Initial contribution
+ *
+ */
+@Component(service = DiscoveryService.class)
+@NonNullByDefault
+public class IRobotDiscoveryService extends AbstractDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(IRobotDiscoveryService.class);
+    private final Runnable scanner;
+    private @Nullable ScheduledFuture<?> backgroundFuture;
+
+    public IRobotDiscoveryService() {
+        super(Collections.singleton(IRobotBindingConstants.THING_TYPE_ROOMBA), 30, true);
+        scanner = createScanner();
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        logger.trace("Starting background discovery");
+        stopBackgroundScan();
+        backgroundFuture = scheduler.scheduleWithFixedDelay(scanner, 0, 60, TimeUnit.SECONDS);
+    }
+
+    @Override
+    protected void stopBackgroundDiscovery() {
+        logger.trace("Stopping background discovery");
+        stopBackgroundScan();
+        super.stopBackgroundDiscovery();
+    }
+
+    private void stopBackgroundScan() {
+        ScheduledFuture<?> scan = backgroundFuture;
+
+        if (scan != null && !scan.isDone()) {
+            scan.cancel(true);
+            backgroundFuture = null;
+        }
+    }
+
+    @Override
+    protected void startScan() {
+        logger.trace("startScan");
+        scheduler.execute(scanner);
+    }
+
+    private Runnable createScanner() {
+        return () -> {
+            long timestampOfLastScan = getTimestampOfLastScan();
+            for (InetAddress broadcastAddress : getBroadcastAddresses()) {
+                logger.debug("Starting broadcast for {}", broadcastAddress.toString());
+
+                DatagramSocket socket;
+                try {
+                    socket = IdentProtocol.sendRequest(broadcastAddress);
+                    while (receivePacketAndDiscover(socket)) {
+                    }
+                } catch (IOException e) {
+                    logger.warn("Error sending broadcast: {}", e.toString());
+                }
+            }
+
+            removeOlderResults(timestampOfLastScan);
+        };
+    }
+
+    private List<InetAddress> getBroadcastAddresses() {
+        ArrayList<InetAddress> addresses = new ArrayList<>();
+
+        for (String broadcastAddress : NetUtil.getAllBroadcastAddresses()) {
+            try {
+                addresses.add(InetAddress.getByName(broadcastAddress));
+            } catch (UnknownHostException e) {
+                logger.warn("Error broadcasting to {}", broadcastAddress, e);
```

Comment:
1. In formatting an exception via {} doesn't dump the full trace on my side, just the message
2. The address is supposed to be raw IP, not a hostname, like 192.168.0.255. Getting UnknownHost on it would be totally strange, some internal system error. I wrote it here only because the IDE forces me to do so.
 I will add the comment in the code.

Reply:
>     1. In formatting an exception via {} doesn't dump the full trace on my side, just the message

The third parameter `e` will lead to dumping the full stack trace.

>     2. The address is supposed to be raw IP, not a hostname, like 192.168.0.255. Getting UnknownHost on it would be totally strange, some internal system error. I wrote it here only because the IDE forces me to do so.
>        I will add the comment in the code.

OK

Reply:
> The third parameter `e` will lead to dumping the full stack trace.
Well, may be i never got this exception so i was just thinking so... Thanks for teaching me some Java :)

Reply:
The third parameter is still there.